### PR TITLE
add orders page

### PR DESCRIPTION
**Pull request by kolyanjy**

### CODE REVIEW

File: app/controllers/orders_controller.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+  def index
+    @orders = Orders::Sort.call(status: params[:sort], user: current_user).orders.decorate
+  end
+
+  def show
+    @order = Order.find_by!(user_id: current_user.id, id: params[:id]).decorate
+  end
+
+  def update
```
Comment (epoberezhny) [inline app/controllers/orders_controller.rb]: add_coupon

File: app/interactors/orders/sort.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
+module Orders
```
Comment (epoberezhny) [inline app/interactors/orders/sort.rb]: Use query object

File: app/views/orders/index.html.haml [hunk 3]
```diff
@@ -0,0 +1,37 @@
+%main.container.general-main-wrap
+  %h1.mt-0.mb-25
+    = t('order.my_orders')
+  %p.lead.small.mb-10.visible-xs Sort by
```
Comment (epoberezhny) [inline app/views/orders/index.html.haml]: I18n

File: app/views/orders/index.html.haml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  .dropdowns.dropdown.general-order-dropdown
+    %a.dropdown-toggle.lead.small{href: "#", 'data-toggle': "dropdown", role: "button", 'aria-haspopup': "true", 'aria-expanded': "false"}
+      - if params[:sort]
+        = t("order.#{params[:sort]}")
```
Comment (epoberezhny) [inline app/views/orders/index.html.haml]: `t(params[:sort], scope: [:order])`

Comment (epoberezhny) [reply app/views/orders/index.html.haml]: params[:sort] == 'Lolkek'

File: app/views/orders/index.html.haml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        = t('order.in_progress')
+      %i.fa.fa-angle-down.dropdown-icon.pull-right
+    %ul.dropdown-menu
+      - Order::FINISH_STATUSES.each do |sort|
+        %li
+          = link_to sort.to_s.split('_').join(' ').capitalize, orders_path(sort: "#{sort}")
+  .visible-xs
+    .general-cart-item.divider-lg-bottom.pt-0
+      %table.table
+        = render 'orders/partials/mobile_orders', orders: @orders
```
Comment (epoberezhny) [inline app/views/orders/index.html.haml]: collection

File: app/controllers/orders_controller.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
+class OrdersController < ApplicationController
+  before_action :authenticate_user!, except: :update
+
+  def index
+    @orders = Orders::Sort.call(status: params[:sort], user: current_user).orders.decorate
+  end
+
+  def show
+    @order = Order.find_by!(user_id: current_user.id, id: params[:id]).decorate
```
Comment (epoberezhny) [inline app/controllers/orders_controller.rb]: `current_user.orders.find_by!(id: params[:id])`

File: spec/interactors/orders/sort_spec.rb [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+  let!(:order_2) { create(:order, :in_delivery_step, user: user) }
+  let!(:order_3) { create(:order, :delivered_step, user: user) }
+  let!(:order_4) { create(:order, :canceled_step, user: user) }
+
+  context 'when status in progress' do
+    let(:status) { :in_progress }
+
+    it do
+      expect(context).to be_success
+      expect(context.orders.first).to eq(order_1)
```
Comment (epoberezhny) [inline spec/interactors/orders/sort_spec.rb]: `eq([order_1])`

**Review by epoberezhny — CHANGES_REQUESTED** [review]

File: app/views/orders/index.html.haml [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  %p.lead.small.mb-10.visible-xs
+    = t('order.sort')
+  .dropdowns.dropdown.general-order-dropdown
+    %a.dropdown-toggle.lead.small{href: "#", 'data-toggle': "dropdown", role: "button", 'aria-haspopup': "true", 'aria-expanded': "false"}
+      = t(@presenter.check_sort(params[:sort]), scope: [:order])
+      %i.fa.fa-angle-down.dropdown-icon.pull-right
+    %ul.dropdown-menu
+      - Order::FINISH_STATUSES.each do |sort|
+        %li
+          = link_to sort.to_s.split('_').join(' ').capitalize, orders_path(sort: "#{sort}")
```
Comment (epoberezhny) [inline app/views/orders/index.html.haml]: 1. I18n
2. why do we need interpolation?

File: app/queries/orders/sort_query.rb [hunk 9]
```diff
@@ -0,0 +1,24 @@
+module Orders
+  class SortQuery
```
Comment (epoberezhny) [inline app/queries/orders/sort_query.rb]: Razve mi chto-to sortiruem tut, Kolyan?

File: app/presenters/orders_presenter.rb [hunk 10]
```diff
@@ -0,0 +1,13 @@
+class OrdersPresenter
+  def check_sort(filter)
```
Comment (epoberezhny) [inline app/presenters/orders_presenter.rb]: sort?

File: app/views/orders/partials/_orders.html.haml [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+  %td
+    = link_to order_path(order) do
+      %span.general-order-number
+        = order.number
+  %td
+    %span.in-grey-900.font-16.fw-300
+      = l(order.created_at, format: :day_month_year)
+  %td
+    %span.font-16.in-grey-900.fw-300
+      = order.status.split('_').join(' ').capitalize
```
Comment (epoberezhny) [inline app/views/orders/partials/_orders.html.haml]: :cry:

File: config/routes.rb [hunk 12]
```diff
@@ -17,18 +17,24 @@
...
   end
+
   resources :order_items, only: %i[update create destroy]
+
   resource :cart, only: :show do
     get :checkout
   end
+
+  resources :orders, only: %i[show index]
+
```
Comment (epoberezhny) [inline config/routes.rb:28]:
```ruby
resources :orders, only: %i[show index] do
  patch :add_coupon
end
```

File: spec/presenters/orders_presenter_spec.rb [hunk 13]
```diff
@@ -0,0 +1,23 @@
+RSpec.describe OrdersPresenter do
+  subject(:presenter) { described_class.new }
+
+  describe 'check_filter' do
+    context 'when filter nil' do
+      let(:filter) { nil }
+
+      it { expect(presenter.check_sort(filter)).to eq(Order::FINISH_STATUSES.first) }
```
Comment (epoberezhny) [inline spec/presenters/orders_presenter_spec.rb]: Use explicit value instead of `Order::FINISH_STATUSES.first`

**Review by epoberezhny — CHANGES_REQUESTED** [review]

File: app/views/orders/partials/_orders.html.haml (hunk 11 above)
Comment (epoberezhny) [reply app/views/orders/partials/_orders.html.haml]: I18n

File: app/queries/orders/sort_query.rb (hunk 9 above)
Comment (dzemlianoi-double) [reply app/queries/orders/sort_query.rb]: A, Kolyan?

Comment (dmbavykin) [reply app/queries/orders/sort_query.rb]: Totally agree with speaker vyshe

File: config/routes.rb [hunk 14]
```diff
@@ -24,9 +24,9 @@
     get :checkout
   end
 
-  resources :orders, only: %i[show index]
-
-  put '/add_couopon', to: 'orders#add_couopon', as: 'add_couopon'
+  resources :orders, only: %i[show index] do
+    patch :add_coupon
```
Comment (epoberezhny) [inline config/routes.rb]: `on: :member`

File: app/views/carts/partials/_coupon.html.haml [hunk 15]
```diff
@@ -1,7 +1,7 @@
 .general-order-wrap
   .row
     .col-sm-4
-      = simple_form_for(Coupon.new, url: add_couopon_path, validate: true, defaults: { input_html: { class: "form-control" }}, html: { method: :put }) do |f|
+      = simple_form_for(Coupon.new, url: order_add_coupon_path(current_order.id), validate: true, defaults: { input_html: { class: "form-control" }}, html: { method: :patch }) do |f|
```
Comment (epoberezhny) [inline app/views/carts/partials/_coupon.html.haml]: do not pass id

File: app/presenters/orders_presenter.rb [hunk 16]
```diff
@@ -1,13 +1,17 @@
 class OrdersPresenter
-  def check_sort(filter)
-    filter_exist?(filter) ? filter : Order::FINISH_STATUSES.first
+  def status_name(status)
```
Comment (epoberezhny) [inline app/presenters/orders_presenter.rb]: rm

File: spec/controllers/orders_controller_spec.rb [hunk 17]
```diff
@@ -36,20 +36,20 @@
   describe '#add_couopon' do
     context 'when coupon exist' do
       let!(:coupon) { create(:coupon) }
-      let(:params) { { coupon: { key: coupon.key }, id: order.id } }
+      let(:params) { { coupon: { key: coupon.key }, order_id: order.id } }
```
Comment (epoberezhny) [inline spec/controllers/orders_controller_spec.rb]: We do we need `order_id` in params?

**Review by epoberezhny — CHANGES_REQUESTED** [review]

**Review by epoberezhny — APPROVED** [review]

**Pull request merged**